### PR TITLE
Added VDS union

### PR DESCRIPTION
**Pull request by tpoterba**
A bunch of people have been asking for this. Once this is in, I'll rebase it against 0.1 and submit another PR. I'll also follow up with a second PR to remove the multiple-VDS-file-reading from `hc.read`.

### CODE REVIEW

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 1]
```diff
@@ -140,6 +140,80 @@ object VariantSampleMatrix {
...
+  def union(datasets: java.util.ArrayList[VariantSampleMatrix[Annotation, Annotation, Annotation]]): VariantSampleMatrix[Annotation, Annotation, Annotation] =
+    union(datasets.asScala.toArray)
+
+  def union(datasets: Array[VariantSampleMatrix[Annotation, Annotation, Annotation]]): VariantSampleMatrix[Annotation, Annotation, Annotation] = {
+    require(datasets.length >= 2)
+
+    checkDatasetSchemasCompatible(datasets)
+    val (first, others) = (datasets.head, datasets.tail)
+    first.copy(rdd = first.sparkContext.union(others.map(_.rdd)).toOrderedRDD(first.rdd.orderedPartitioner.kOk, first.rdd.vct))
+  }
```
Comment (cseed) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: This loses the first dataset?

File: python/hail/dataset.py [hunk 2]
```diff
@@ -2679,6 +2679,59 @@ def join(self, right):
...
+
+        This method can trigger a shuffle, if partitions from two datasets overlap.
+
+        :param vds_type: Datasets with which to take union.
+        :type vds_type: tuple of :class:`.VariantDataset`
+
+        :return: Dataset with variants from all datasets.
+        :rtype: :class:`.VariantDataset`
+        """
+        if len(datasets) < 2:
```
Comment (cseed) [inline python/hail/dataset.py]: Support 1 dataset.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 3]
```diff
@@ -140,6 +140,80 @@ object VariantSampleMatrix {
       tGen = (t: Type, v: Annotation) => t.genValue.resize(20),
       makeKOk = (vSig: Type) => vSig.orderedKey)
       .gen(hc)
+
+  def checkDatasetSchemasCompatible(datasets: Array[VariantSampleMatrix[Annotation, Annotation, Annotation]]) {
```
Comment (cseed) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: This is a duplicate of the routine of the same name in HailContext that's used for read.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 1 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: Huh, yeah. How does this pass? My tests must be wrong.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 3 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I'm going to remove that in the next PR. I could consolidate that here, is that preferred?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 1 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: Yep, my test was bad and I should feel bad.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 3 above)
Comment (tpoterba) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: In particular, the error message is much better in the one used for read. That's why I left it (and would prefer to leave it duplicated in 0.1)

File: python/hail/dataset.py [hunk 4]
```diff
@@ -2679,6 +2679,61 @@ def join(self, right):
...
+        >>> vds_union1 = vds_autosomal.union(vds_chromX, vds_chromY)
+        >>> vds_union2 = all_vds[0].union(*all_vds[1:])
+        >>> vds_union3 = VariantDataset.union(*all_vds)
+
+        **Notes**
+
+        In order to combine two datasets, four requirements must be met:
+         - the column keys must be identical (both type and identity, field order within structs matters).
+         - the row key and annotation schemas must match (field order within structs matters).
+         - the row (variant) annotation schemas must match (field order within structs matters).
```
Comment (catoverdrive) [inline python/hail/dataset.py]: is the third requirement not a suset of the second one? Or am I misreading what the second requirement is?

Comment (tpoterba) [reply python/hail/dataset.py]: oops, good catch

**Comment by tpoterba** [comment]
yeah, my "test setup" before the doc tests was wrong (I used "vds1" instead of "vds")

**Pull request merged**